File: lint-arwaky/archive/infrastructure/python_ruff_adapter.py
"""Ruff adapter for Python linting."""

from __future__ import annotations

from ..taxonomy import (
    AdapterName,
    ColumnNumber,
    ComplianceStatus,
    ErrorCode,
    FilePath,
    LineNumber,
    LintMessage,
    LintResult,
    LintResultList,
    Severity,
    PatternList,
    Timeout,
    AdapterError,
    ScanError,
    ErrorMessage,
)

import json
import logging
import os

from ..contract import ILinterAdapterPort, ICommandExecutorPort, IPathNormalizationPort

logger = logging.getLogger("mcp.adapters.ruff")


class RuffAdapter(ILinterAdapterPort):
    """Adapter for Ruff linter."""

    def __init__(
        self,
        executor: ICommandExecutorPort,
        path_norm: IPathNormalizationPort,
        bin_path: FilePath | None = None,
    ):
        self.executor = executor
        self.path_norm = path_norm
        self.bin_path = bin_path

    def name(self) -> AdapterName:
        return AdapterName(value="ruff")

    def _resolve_working_dir(self, path: FilePath) -> FilePath:
        path_str = str(path)
        try:
            from pathlib import Path
            current = Path(os.path.abspath(path_str))
            if current.is_file():
                current = current.parent
            for _ in range(10):
                if (
                    (current / "auto_linter.config.python.yaml").is_file() or (current / "auto_linter.config.javascript.yaml").is_file() or (current / "auto_linter.config.rust.yaml").is_file()
                    or (current / "auto_linter.config.python.yaml").is_file()
                    or (current / "pyproject.toml").is_file()
                    or (current / ".git").is_dir()
                ):
                    return FilePath(value=str(current))
                if current.parent == current:
                    break
                current = current.parent
        except Exception as e:
            logger.debug(f"Error resolving working directory: {e}")
        return FilePath(value=".")

    async def scan(self, path: FilePath) -> LintResultList | ScanError | AdapterError:
        path_str = str(path)
        results = []
        try:
            abs_path = os.path.abspath(path_str)
            executable = self._resolve_executable(AdapterName(value="ruff"))

            cmd = [
                str(executable),
                "check",
                abs_path,
                "--output-format=json",
                "--exit-zero",
                "--no-cache",
            ]
            result = await self.executor.execute_command(
                command=PatternList(values=cmd),
                working_dir=self._resolve_working_dir(path),
                timeout=Timeout(value=60.0),
            )

            if result.stderr:
                logger.debug(f"Ruff stderr: {str(result.stderr)}")

            stdout_str = str(result.stdout).strip()
            findings = json.loads(stdout_str) if stdout_str else []
            for f in findings:
                filename_vo = self.path_norm.normalize_path(
                    FilePath(value=f.get("filename", ""))
                )
                filename = str(filename_vo)
                if not filename:
                    continue

                results.append(
                    LintResult(
                        file=FilePath(value=filename),
                        line=LineNumber(value=f.get("location", {}).get("row", 0)),
                        column=ColumnNumber(
                            value=f.get("location", {}).get("column", 0)
                        ),
                        code=ErrorCode(code=f.get("code", "UNKNOWN")),
                        message=LintMessage(value=f.get("message", "")),
                        source=AdapterName(value="ruff"),
                        severity=self._map_severity(
                            f.get("severity", ""), f.get("code", "")
                        ),
                    )
                )
        except Exception as e:
            logger.error(f"Ruff scan failed: {e}")
            return ScanError(
                path=path,
                message=ErrorMessage(value=str(e)),
                adapter_name=self.name()
            )

        return LintResultList(values=results)

    async def apply_fix(self, path: FilePath) -> ComplianceStatus | AdapterError:
        path_str = str(path)
        try:
            executable = self._resolve_executable(AdapterName(value="ruff"))
            cmd = [str(executable), "check", path_str, "--fix", "--exit-zero"]
            await self.executor.execute_command(
                command=PatternList(values=cmd),
                working_dir=self._resolve_working_dir(path),
                timeout=Timeout(value=60.0),
            )
            return ComplianceStatus(value=True)
        except Exception as e:
            logger.error(f"Error applying Ruff fixes: {e}")
            return AdapterError(
                adapter_name=self.name(),
                message=ErrorMessage(value=str(e))
            )

    def _resolve_executable(self, name: AdapterName) -> FilePath:
        if self.bin_path:
            return FilePath(value=os.path.join(str(self.bin_path), str(name)))
        return FilePath(value=str(name))

    def _map_severity(self, raw, code="") -> Severity:
        """Map Ruff severity string to Severity VO."""
        s = str(raw).upper()
        c = str(code).upper()

        # Logic errors and syntax errors are CRITICAL
        if c.startswith(("F", "E9")):
            return Severity.CRITICAL

        if "ERROR" in s:
            return Severity.HIGH
        if "WARNING" in s:
            return Severity.MEDIUM
        return Severity.LOW

    async def _to_lint_result(self, item, scan_path: FilePath) -> LintResult:
        data = item
        code = data.get("code", "")
        severity = Severity.MEDIUM
        if code.startswith("F") or code.startswith("E9"):
            severity = Severity.HIGH
        elif code.startswith("W"):
            severity = Severity.LOW

        filename_vo = self.path_norm.resolve_infrastructure_path(
            FilePath(value=data["filename"]), scan_path
        )

        return LintResult(
            file=filename_vo,
            line=LineNumber(value=data["location"]["row"]),
            column=ColumnNumber(value=data["location"]["column"]),
            code=ErrorCode(code=code),
            message=LintMessage(value=data["message"]),
            source=self.name(),
            severity=severity,
        )

File: lint-arwaky/src-rust/infrastructure/python_ruff_adapter.rs
/// Ruff adapter for Python linting.
use crate::contract::{ILinterAdapterPort, ICommandExecutorPort, IPathNormalizationPort, LinterError};
use crate::taxonomy::*;
use async_trait::async_trait;
use serde_json::Value;
use std::sync::Arc;

pub struct RuffAdapter {
    executor: Arc<dyn ICommandExecutorPort>,
    path_norm: Arc<dyn IPathNormalizationPort>,
    bin_path: Option<FilePath>,
}

impl RuffAdapter {
    pub fn new(
        executor: Arc<dyn ICommandExecutorPort>,
        path_norm: Arc<dyn IPathNormalizationPort>,
        bin_path: Option<FilePath>,
    ) -> Self {
        Self {
            executor,
            path_norm,
            bin_path,
        }
    }

    fn resolve_executable(&self) -> String {
        self.bin_path
            .as_ref()
            .map(|p| p.value.clone())
            .unwrap_or_else(|| "ruff".to_string())
    }

    fn map_severity(&self, severity: &str, _code: &str) -> Severity {
        match severity {
            "error" => Severity::ERROR,
            "warning" => Severity::WARNING,
            "info" => Severity::INFO,
            _ => Severity::WARNING,
        }
    }
}

#[async_trait]
impl ILinterAdapterPort for RuffAdapter {
    fn name(&self) -> AdapterName {
        AdapterName::new("ruff")
    }

    async fn scan(&self, path: &FilePath) -> Result<LintResultList, LinterError> {
        let executable = self.resolve_executable();
        let cmd = vec![
            executable,
            "check".to_string(),
            path.value.clone(),
            "--output-format=json".to_string(),
            "--exit-zero".to_string(),
            "--no-cache".to_string(),
        ];

        let command = PatternList::new(cmd);
        let working_dir = FilePath::new("."); // Simplified

        match self.executor.execute_command(command, working_dir, Some(std::time::Duration::from_secs(60))).await {
            Ok(response) => {
                let stdout = response.stdout.value;
                let findings: Vec<Value> = serde_json::from_str(&stdout).unwrap_or_default();
                let mut results = Vec::new();

                for f in findings {
                    let filename = f.get("filename").and_then(|v| v.as_str()).unwrap_or("");
                    let row = f.get("location").and_then(|l| l.get("row")).and_then(|v| v.as_i64()).unwrap_or(0);
                    let col = f.get("location").and_then(|l| l.get("column")).and_then(|v| v.as_i64()).unwrap_or(0);
                    let code = f.get("code").and_then(|v| v.as_str()).unwrap_or("UNKNOWN");
                    let message = f.get("message").and_then(|v| v.as_str()).unwrap_or("");
                    let severity_str = f.get("severity").and_then(|v| v.as_str()).unwrap_or("");

                    results.push(LintResult {
                        file: FilePath::new(filename),
                        line: LineNumber::new(row),
                        column: ColumnNumber::new(col),
                        code: ErrorCode::new(code),
                        message: LintMessage::new(message),
                        source: Some(self.name()),
                        severity: self.map_severity(severity_str, code),
                        ..Default::default()
                    });
                }
                Ok(LintResultList::new(results))
            }
            Err(e) => Err(LinterError::Adapter(AdapterError {
                message: ErrorMessage::new(format!("Ruff execution failed: {}", e)),
                ..Default::default()
            })),
        }
    }

    async fn apply_fix(&self, path: &FilePath) -> Result<ComplianceStatus, LinterError> {
        let executable = self.resolve_executable();
        let cmd = vec![
            executable,
            "check".to_string(),
            path.value.clone(),
            "--fix".to_string(),
            "--exit-zero".to_string(),
        ];

        let command = PatternList::new(cmd);
        let working_dir = FilePath::new(".");

        match self.executor.execute_command(command, working_dir, Some(std::time::Duration::from_secs(60))).await {
            Ok(_) => Ok(ComplianceStatus::new(true)),
            Err(e) => Err(LinterError::Adapter(AdapterError {
                message: ErrorMessage::new(format!("Ruff fix failed: {}", e)),
                ..Default::default()
            })),
        }
    }
}
